solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -2899,6 +2920,23 @@ public NotInClusterStateException(ErrorCode code, String msg) {
     }
   }
 
+  /**
+   * Thrown during pre-register process if the replica is found in clusterstate but is inconsistent
+   * with current node name
+   */
+  public static class InconsistentClusterStateException extends SolrException {
```

Comment:
Why a new Exception? The `replica` field is not used anywhere

Reply:
The new exception better identifies a new type of exception. This cannot be `NotInClusterStateException`, otherwise the `CoreContainer` will attempt to unload (ie delete)  the core at https://github.com/fullstorydev/lucene-solr/blob/release/8.8/solr/core/src/java/org/apache/solr/core/CoreContainer.java#L1525.

The extra field is just to store the info which could be useful for any future handling/logging

Reply:
My concern is that this Exception is not caught anywhere or we do not do an `instanceof` check

Reply:
Since this is a non runtime exception, the caller should be forced to handle such exception specifically or this should fall to any handling of unexpected exceptions of existing logic? :)

In this case, the caller `preRegister` does already have handling for general unexpected exception, so this is supposed to work as is - be wrapped at `SolrException`, re-thrown and caught in `createFromDescriptor`, but unlike `NotInClusterStateException`, which unload(delete) the physical directory, it will just interrupt the core loading, wrapped and thrown again, and finally be logged in `CoreContainer#load` without affecting other cores.

Such described flow is already the handling of any unexpected exception which I find quite fitting for our scenario. And there's an added test case to ensure a single `InconsistentClusterStateException ` should not affect loading of other "healthy" cores. 

Lemme know if you have any thoughts on different handling!